logger/logger.go

```diff
@@ -0,0 +1,137 @@
+package logger
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"sync"
+	"time"
+)
+
+type Level int
+
+const (
+	INVALID Level = iota
+	TRACE
+	DEBUG
+	INFO
+	WARN
+	ERROR
+	FATAL
+)
+
+func (l Level) String() string {
+	switch l {
+	case INVALID:
+		return "INVALID"
+	case TRACE:
+		return "TRACE"
+	case DEBUG:
+		return "DEBUG"
+	case INFO:
+		return "INFO"
+	case WARN:
+		return "WARN"
+	case ERROR:
+		return "ERROR"
+	case FATAL:
+		return "FATAL"
+	default:
+		return ""
+	}
+}
+
+func (l Level) MarshalJSON() ([]byte, error) {
+	return json.Marshal(l.String())
+}
+
+type key int
+
+const (
+	level key = iota
+	values
+)
+
+type Pair struct {
+	Key   string
+	Value interface{}
+}
+
+type Formatter interface {
+	Log(vals ...interface{})
+}
+
+type FormatterFunc func(vals ...interface{})
+
+func (ff FormatterFunc) Log(vals ...interface{}) {
+	ff(vals...)
+}
+
+var impl Formatter = DefaultLogger{os.Stdout}
+var o sync.Once
+
+func Config(i Formatter) {
+	o.Do(func() {
```

Comment:
Is it OK that the first call to Config() will success and subsequent calls will fail? Seems like there's no feedback on subsequent calls.

Reply:
You could have a point here. It makes sense when used as a single global logger but maybe a warning of some sort when called multiple times?

Reply:
Good point. I'm trying to avoid synchronization, but it might be impossible. I'm debating what the least worst API looks like. I don't think people expect to check error on a Config call, so I'm not keen on that. That leaves making every call to Config work, which also means that I should do a sync.Lock on each read and write of the level (to avoid hypothetical publishing synchronization errors; this is one of the times where it'd be nice if Go had volatile).